File: WoofAndroid/testutils/src/main/kotlin/com/perrystreet/woof/testutils/scheduler/TimeAdvancingFactory.kt
package com.perrystreet.woof.testutils.scheduler

import org.koin.test.KoinTest
import org.koin.test.inject
import java.util.concurrent.TimeUnit

class TimeAdvancingFactory : KoinTest {
    private val scheduler: TestSchedulerProvider by inject()

    fun withTestSchedulers() = apply {
        scheduler.useTestSchedulers()
    }

    fun tick() {
        scheduler.testScheduler.advanceTimeBy(1, TimeUnit.MICROSECONDS)
    }

    fun advanceMillis(amount: Long) {
        scheduler.testScheduler.advanceTimeBy(amount, TimeUnit.MILLISECONDS)
    }
}

File: WoofIOS/SwiftPackages/TestUtils/Sources/TestUtils/Scheduler/TimeAdvancingFactory.swift
import Foundation
import Swinject
import SwinjectAutoregistration

public final class TimeAdvancingFactory {
    private let scheduler: TestSchedulerProvider

    public init(_ container: Container) {
        scheduler = container~>
    }

    @discardableResult
    public func withTestSchedulers() -> Self {
        scheduler.useTestSchedulers()
        return self
    }

    public func tick() {
        scheduler.testScheduler.advance(by: .microseconds(1))
    }

    public func advanceMillis(_ amount: Int) {
        scheduler.testScheduler.advance(by: .milliseconds(amount))
    }
}
